src/stan_math_backend/Transform_Mir.ml

```diff
@@ -395,6 +395,65 @@ let%expect_test "insert before" =
   [%sexp (l : int list)] |> print_s ;
   [%expect {| (1 2 3 4 5 999 6) |}]
 
+let rec add_validate_dims outvars stmts =
+  let params =
+    List.filter_map
+      ~f:(function
+        | decl_id, Program.({out_block= Parameters; out_trans; _}) -> (
+          match out_trans with
+          | Program.Simplex ->
+              Some (decl_id, Internal_fun.FnValidateSizeSimplex)
+          | Program.UnitVector -> Some (decl_id, FnValidateSizeUnitVector)
+          | _ -> None )
+        | _ -> None)
+      outvars
+    |> String.Map.of_alist_exn
+  in
+  let fnvalidate id =
+    Option.value ~default:Internal_fun.FnValidateSize (Map.find params id)
+  in
+  let f = function
+    | Stmt.Fixed.({pattern= Decl {decl_id; decl_type= Sized st; _}; meta}) as
+      decl ->
+        let check fn x =
+          Stmt.Fixed.
+            { meta
+            ; pattern=
+                NRFunApp
```

Comment:
You can use the `internal_funapp` helper here to make this a little more readable, hopefully.